qemu/tests/qemu_guest_agent.py

```diff
@@ -785,16 +782,39 @@ def _action_after_fsfreeze(self, *args):
             return
 
         if not self._open_session_list:
-            raise error.TestError("Could not find any opened session")
+            self.test.error("Could not find any opened session")
         # Use the last opened session to send cmd.
         session = self._open_session_list[-1]
         try:
             session.cmd(self.params["gagent_fs_test_cmd"])
         except aexpect.ShellTimeoutError:
             logging.debug("FS freeze successfully.")
         else:
-            raise error.TestFail("FS freeze failed, guest still can"
-                                 " write file")
+            self.test.fail("FS freeze failed, guest still can"
+                           " write file")
+
+    def gagent_check_reset_reboot_shutdown(self, test, params, env):
+
+        """
+        Execute "system_reset" and send "shutdown,reboot"
+        command to guest agent after FS freezed
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment.
+        """
+        vm = env.get_vm(params["main_vm"])
+        vm.verify_alive()
+        gagent = self.gagent
+        gagent.fsfreeze()
+
+        if gagent.shutdown(gagent.SHUTDOWN_MODE_POWERDOWN):
+            gagent.verify_fsfreeze_status(gagent.FSFREEZE_STATUS_FROZEN)
+
+        if gagent.shutdown(gagent.SHUTDOWN_MODE_REBOOT):
+            gagent.verify_fsfreeze_status(gagent.FSFREEZE_STATUS_FROZEN)
```

Comment:
If the guest has been shut down, how does it be rebooted at this point?

Reply:
@luckyh  when send agent shutdown command to guest.  guest file system is frozen status(L808:gagent.fsfreeze()). So guest isn't really closed.  If guest is closed, then it will be a production bug.   Another it will use vm.monitor.cmd("system_reset") to thaw guest file system.